File: python/array/ContainerWithMostWater.py
class Solution:
    # @return an integer
    def maxArea(self, height):
        low = 0
        high = len(height) - 1
        maxWater = 0
        while low < high:
            currLength = high - low
            currHeight = min(height[high], height[low])
            maxWater = max(currLength*currHeight, maxWater)
            if height[low] < height[low+1]:
                low += 1
            elif height[high] < height[high-1]:
                high -= 1
            else:
                if height[low] > height[high]:
                    high -= 1
                else:
                    low += 1
        return maxWater

if __name__ == "__main__":
    solution = Solution()

    print solution.maxArea([5, 2, 42, 1, 5, 30, 4, 11, 9, 4])

File: cpp/array/ContainerWithMostWater.cpp
class Solution {
public:
    int maxArea(vector<int> &height) {
        int low = 0;
        int high = height.size() - 1;
        int max_water = 0;
        while (low < high) {
            int curr_length = high - low;
            int curr_height = std::min(height[low], height[high]);
            max_water = std::max(curr_length * curr_height, max_water);
            if (height[low] < height[low+1]) {
                low ++;
            } else if (height[high] < height[high-1]) {
                high --;
            } else {
                if (height[low] < height[high]) {
                    low ++;
                } else {
                    high --;
                }
            }
        }
        return max_water;
    }
};
